pkg/runtime/nodejs/nodejs.go

```diff
@@ -0,0 +1,240 @@
+	for _, m := range maps {
+		if pathname := m.Pathname; pathname != "" {
+			if m.Perms.Execute {
+				if isNodeJSLib(pathname) {
+					found = true
+					lib = pathname
+					break
+				}
+			}
+		}
+	}
+	if !found {
+		// If we didn't find a library, we can't be sure that this is a nodejs process.
+		return false, nil
+	}
+
+	ef, err := elf.Open(absolutePath(proc, lib))
+	if err != nil {
+		return false, fmt.Errorf("open elf file: %w", err)
+	}
+
+	isNodeJS, err = runtime.HasSymbols(ef, nodejsIdentifyingSymbols)
+	if err != nil {
+		return false, fmt.Errorf("failed to check for symbols: %w", err)
+	}
+
+	return isNodeJS, nil
+}
+
+func RuntimeInfo(proc procfs.Proc) (*runtime.Runtime, error) {
+	isNodeJS, err := IsRuntime(proc)
+	if err != nil {
+		return nil, fmt.Errorf("is runtime: %w", err)
+	}
+	if !isNodeJS {
+		return nil, nil //nolint:nilnil
+	}
+
+	rt := &runtime.Runtime{
+		Type: "nodejs",
+	}
+
+	exe, err := proc.Executable()
+	if err != nil {
+		return rt, err
+	}
+
+	f, err := os.Open(absolutePath(proc, exe))
+	if err != nil {
+		return rt, fmt.Errorf("open executable: %w", err)
+	}
+	defer f.Close()
+
+	versionString, err := versionFromData(f)
+	if err == nil {
+		version, err := semver.NewVersion(versionString)
+		if err != nil {
+			return rt, fmt.Errorf("new version, %s: %w", versionString, err)
+		}
+		rt.Version = version
+		return rt, nil
+	}
+
+	maps, err := proc.ProcMaps()
+	if err != nil {
+		return nil, fmt.Errorf("error reading process maps: %w", err)
+	}
+
+	var (
+		found bool
+		lib   string
+	)
+	for _, m := range maps {
+		if pathname := m.Pathname; pathname != "" {
+			if m.Perms.Execute {
+				if isNodeJSLib(pathname) {
+					found = true
+					lib = pathname
+					break
+				}
+			}
+		}
+	}
+	if !found {
+		return rt, fmt.Errorf("library %q not found in process maps", exe)
+	}
+
+	lf, err := os.Open(absolutePath(proc, lib))
+	if err != nil {
+		return rt, fmt.Errorf("open library: %w", err)
+	}
+	defer lf.Close()
+
+	versionString, err = versionFromData(lf)
+	if err != nil {
+		return rt, fmt.Errorf("version from data: %w", err)
+	}
+
+	version, err := semver.NewVersion(versionString)
+	if err != nil {
+		return rt, fmt.Errorf("new version, %s: %w", versionString, err)
+	}
+	rt.Version = version
+	return rt, nil
+}
+
+func isNodeJSBin(exe string) bool {
+	return path.Base(exe) == "node" || path.Base(exe) == "nodemon" || path.Base(exe) == "nodejs"
+}
+
+var libRegex = regexp.MustCompile(`/libnode.so\.[0-9]+`)
+
+func isNodeJSLib(lib string) bool {
+	return libRegex.MatchString(lib)
+}
+
+func absolutePath(proc procfs.Proc, p string) string {
+	return path.Join("/proc/", strconv.Itoa(proc.PID), "/root/", p)
+}
+
+func versionFromData(f *os.File) (string, error) {
+	ef, err := elf.NewFile(f)
+	if err != nil {
+		return "", fmt.Errorf("new file: %w", err)
+	}
+	defer ef.Close()
+
+	var lastError error
+	for _, sec := range ef.Sections {
+		if sec.Name == ".data" || sec.Name == ".rodata" {
+			data, err := sec.Data()
```

Comment:
perf nit: This might allocate a new buffer, I don't expect it to be very large, but might be worth checking some node js executables

Reply:
You are spot on I have already an optimization lined up to optimize this part. Stay tuned ;)

Reply:
@javierhonduco https://github.com/parca-dev/parca-agent/pull/2254